extras/test_hull_performace.cpp

```diff
@@ -0,0 +1,175 @@
+#include "meshIO.h"
+#include "samples.h"
+using namespace std;
+using namespace glm;
+using namespace manifold;
+
+// Epick = Exact predicates Inexact constructions. Seems fair to use to compare
+// to Manifold, which seems to store double coordinates.
+typedef CGAL::Epick Kernel;
+
+// Epeck = Exact predicates Exact constructions. What OpenSCAD uses to guarantee
+// geometry ends up where it should even after many operations. typedef
+// CGAL::Epeck Kernel;
+
+typedef CGAL::Point_3<Kernel> Point;
+typedef CGAL::Surface_mesh<Point> TriangleMesh;
+typedef CGAL::SM_Vertex_index Vertex;
+
+// Converts Manfiold to CGAL Surface Mesh
+void manifoldToCGALSurfaceMesh(Manifold &manifold, TriangleMesh &cgalMesh) {
+  auto maniMesh = manifold.GetMesh();
+
+  const size_t n = maniMesh.vertPos.size();
+  std::vector<Vertex> vertices(n);
+  for (size_t i = 0; i < n; i++) {
+    auto &vert = maniMesh.vertPos[i];
+    vertices[i] = cgalMesh.add_vertex(Point(vert.x, vert.y, vert.z));
+  }
+
+  for (auto &triVert : maniMesh.triVerts) {
+    std::vector<Vertex> polygon{vertices[triVert[0]], vertices[triVert[1]],
+                                vertices[triVert[2]]};
+    cgalMesh.add_face(polygon);
+  }
+}
+
+// Prints the volume and area of the manifold and the convex hull of our
+// implementation,
+void PrintVolArea(Manifold &manfiold_obj, Manifold &hullManifold) {
+  std::cout << manfiold_obj.GetProperties().volume << ",";
+  std::cout << hullManifold.GetProperties().volume << ",";
+  std::cout << manfiold_obj.GetProperties().surfaceArea << ",";
+  std::cout << hullManifold.GetProperties().surfaceArea << ",";
+  std::cout << manfiold_obj.NumTri() << ",";
+  std::cout << hullManifold.NumTri() << ",";
+  return;
+}
+
+// Prints the volume and area of the manifold and Convex Hull of CGAL
+// Implementation
+void PrintVolAreaCGAL(TriangleMesh &cgalMesh, TriangleMesh &cgalHull) {
+  std::cout << CGAL::Polygon_mesh_processing::volume(cgalMesh) << ",";
+  std::cout << CGAL::Polygon_mesh_processing::volume(cgalHull) << ",";
+  std::cout << CGAL::Polygon_mesh_processing::area(cgalMesh) << ",";
+  std::cout << CGAL::Polygon_mesh_processing::area(cgalHull) << ",";
+  std::cout << std::distance(cgalMesh.faces_begin(), cgalMesh.faces_end())
+            << ",";
+  std::cout << std::distance(cgalHull.faces_begin(), cgalHull.faces_end())
+            << ",";
+  return;
+}
+
+void PrintManifold(Manifold input, Manifold (Manifold::*hull_func)() const) {
+  auto start = std::chrono::high_resolution_clock::now();
+  Manifold hullManifold = (input.*hull_func)();
+#ifdef MANIFOLD_EXPORT
+  if (options.exportModels) {
+    ExportMesh("testHullManifold.glb", hullManifold.GetMesh(), {});
+  }
+#endif
+  auto end = std::chrono::high_resolution_clock::now();
+  PrintVolArea(input, hullManifold);
+  std::chrono::duration<double> elapsed = end - start;
+  std::cout << elapsed.count() << " sec";
+}
+
+void PrintCGAL(Manifold input, Manifold (Manifold::*hull_func)() const) {
```

Comment:
We probably don't need a separate CGAL one as well.

Reply:
The reason, I kept it, is since we had discussed that we don't want to keep the other implementations in our code. CGAL, seemed like the simplest way to compare the outputs relatively.

Reply:
I mean, we can just use the function for `PrintManifold`? Just run CGAL hull in the `hull_func` function

Reply:
Right, since we don't need the Manfiold Method function, we can just use that. I will change that too. Now, that I think of it, that was a very redundant piece of code thanks for pointing it out.

Reply:
Ok so, since we aren't storing the output of CGAL as a manifold, I am just computing the CGAL output, and then printing the outputs using the CGAL functions. I didn't want to convert the CGAL output back to manifold as that would add time overhead, so what should I do in this case?

Reply:
The simplest way would be to do the time measurement in the callback, or you can split the function into two, the actual hull operation as well as a postprocessing operation which returns a manifold, and you only measure the time in the actual hull operation.

```cpp
class HullImpl {
  public:
    // actual hull operation, we will measure the time needed to evaluate this function.
    virtual void hull(const std::vector<glm::vec3>& points);
    // return the result of the hull operation, should be called after `hull`.
    virtual manifold::Manifold result();
};
```

And just let the implementations inherit this class. (or if you use template, you don't need inheritance).